js/memory.js

```diff
@@ -1,19 +1,35 @@
 class MemoryGame {
   constructor(cards) {
     this.cards = cards;
+    this.pickedCards = [];
+    this.pairsClicked = 0;
+    this.pairsGuessed = 0;
     // add the rest of the class properties here
   }
 
   shuffleCards() {
-    // ... write your code here
+    if (!this.cards) return undefined;
+    //Fisher-Yates
+    for (let i = this.cards.length - 1; i > 0; i--) {
+      let j = Math.floor(Math.random() * (i + 1));
+      let k = this.cards[i];
+      this.cards[i] = this.cards[j];
+      this.cards[j] = k;
+    }
+    return this.cards;
   }
 
   checkIfPair(card1, card2) {
-    // ... write your code here
+    this.pairsClicked++;
+    if (card1 === card2) {
+      this.pairsGuessed++;
+      return true;
+    }
```

Comment:
I leave you a suggestion that I wrote just know (so it may not work, but you should get the point)


```
class MemoryGame {
  constructor(cards) {
    this.cards = cards;
    this.pickedCards = [];
    this.pairsClicked = 0;
    this.pairsGuessed = 0;
  }

  pickCard(card) {
    this.pickedCards.push(card);
  }

  playRound(selectedCard) {
    const hasPickedTwo = this.pickedCards.length >= 2;
    if (!hasPickedTwo) {
      selectedCard.classList.toggle('turned');
      this.pickCard(selectedCard);
      return;
    }

    console.log('select just two cards');
  }

  flipCards() {
    this.pickedCards.forEach((card) => card.classList.toggle('turned'));
  }

  freezeCards() {
    this.pickedCards.forEach((card) => (card.style.pointerEvents = 'none'));
  }

  shuffleCards() {
    this.cards.sort(() => 0.5 - Math.random());
  }

  checkIfPair(card1, card2) {
    this.pairsClicked++;
    if (card1 === card2) {
      this.freezeCards();
      this.pickedCards = [];
      this.pairsGuessed++;
      return true;
    }

    return false;
  }

  checkIfFinished() {
    return this.pairsGuessed >= 2;
  }
}

const cards = document.querySelector('.card');
const memoryGame = new MemoryGame(cards);

function updateClicks() {
  const pairsClicked = document.querySelector('#pairs-clicked');
  pairsClicked.innerHTML = memoryGame.pairsClicked;
}

function updateGuessed() {
  const pairsGuessed = document.querySelector('#pairs-guessed');
  pairsGuessed.innerHTML = memoryGame.pairsClicked;
}

function cardName(card) {
  return card.getAttribute('data-card-name');
}

function onCardClick({ currentTarget: clickedCard }) {
  memoryGame.playRound(clickedCard);
  updateClicks();
  const hasSelectedTwoCards = memoryGame.pickedCards.length === 2;
  if (hasSelectedTwoCards) {
    const [card1, card2] = memoryGame.pickedCards;
    memoryGame.checkIfPair(cardName(card1), cardName(card2));
    updateGuessed();
    if (memoryGame.checkIfFinished()) {
      alert('you won');
    }
  }
}

cards.forEach((card) => card.addEventListener('click', onCardClick));

```